### ISSUE TITLE
`use-case-hack` is a property of a store, not a global setting (except for `libnixfile`-level commands)

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#9318]
**Describe the bug**

```console
$ nix copy --from ~/tmp-store-normal --to ~/tmp-store-hacked?use-case-hack=true /nix/store/6hypxafvp2fcadphppjcxszx0qcahqbz-dummy
warning: unknown setting 'use-case-hack'
```

I shouldn't have to apply the case hack (unhack) to the `--from` store.

**Steps To Reproduce**

See above.

**Expected behavior**

- The case hack is a store setting.
- Ideally, need for the case hack is autodetected based on the file system where the store resides
   - #1446

**`nix-env --version` output**

**Additional context**

`libnixfile` isn't a thing (yet?), but is a useful indictor of the abstraction level: Nix operations on files without the need for an actual store.

- #1446

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by edolstra** [comment]
Do we even need the case hack anymore? IIRC it was added once upon a time to support NixOps on macOS, where we ran into Linux kernel headers having case conflicts. 

Also, since our installer creates a volume for the Nix store these days, I guess we can just enable case sensitivity?

**Comment by roberth** [comment]
Having a case sensitive store on macOS sounds nice, but it's not free of side effects.

- Build impurity: this difference will be observable in the sandbox, and outputs will differ.
- It doesn't remove the problem, but rather moves it from the _remote store <-> local store_ boundary to the _local store <-> rest of system_ boundary, which seems like a larger surface area in practical use.

If we make such a change we better be sure that it's not a regression in terms of actual use, so I wouldn't block improvements to the case hack functionality on this decision. Note also that we'll have to support the case hack until everyone's had a good chance to migrate, and macOS doesn't have a monopoly on case insensitive file systems.

**Comment by j-baker** [comment]
I don't believe this hack can be removed. Among other things, the Nix installer on MacOS creates the Nix store on a case-insensitive volume.
